src/compute/src/logging/compute.rs

```diff
@@ -609,32 +609,33 @@ where
         import_id: GlobalId,
         export_ids: Vec<GlobalId>,
     ) -> Self {
-        let mut previous_time = None;
+        // Prepare retraction events as `LogOnDrop`s, to ensure they are logged even when the
+        // dataflow is dropped before the input advances to the empty frontier.
+        let mut retractions = Vec::<LogOnDrop>::with_capacity(export_ids.len());
+
         self.inspect_container(move |event| {
-            if let Err(frontier) = event {
-                if let Some(previous) = previous_time {
-                    for &export_id in export_ids.iter() {
-                        logger.log(ComputeEvent::ImportFrontier {
-                            import_id,
-                            export_id,
-                            time: previous,
-                            diff: -1,
-                        });
-                    }
-                }
-                if let Some(time) = frontier.get(0) {
-                    for &export_id in export_ids.iter() {
-                        logger.log(ComputeEvent::ImportFrontier {
-                            import_id,
-                            export_id,
-                            time: *time,
-                            diff: 1,
-                        });
-                    }
-                    previous_time = Some(*time);
-                } else {
-                    previous_time = None;
-                }
+            let Err(frontier) = event else { return };
+
+            // Log frontier retractions by dropping them.
+            retractions.drain(..).for_each(drop);
+
+            let Some(time) = frontier.get(0) else { return };
+            for &export_id in export_ids.iter() {
+                logger.log(ComputeEvent::ImportFrontier {
+                    import_id,
+                    export_id,
+                    time: *time,
+                    diff: 1,
+                });
+                retractions.push(LogOnDrop::new(
+                    logger.clone(),
+                    ComputeEvent::ImportFrontier {
+                        import_id,
+                        export_id,
+                        time: *time,
+                        diff: -1,
+                    },
+                ));
```

Comment:
I think we can achieve the same with less cloning. Cloning the logger isn't expensive, but it's also not free. We could have the `LogOnDrop` wrap the `retractions`, so we would only need to do the work to maintain the time at which to drop, since `Vec<export_id>`/`import_id` stay constant.

This would mean that we still need two places where the retraction is logged, once in the `drop` case, and once in the regular case as it is done now.

Reply:
Yes, that sounds like it will work!